### PR TITLE
service config: default service config

### PR DESCRIPTION
**Pull request by lyuxuan**

### CODE REVIEW

File: dialoptions.go [hunk 1]
```diff
@@ -446,6 +448,22 @@ func WithDisableServiceConfig() DialOption {
 	})
 }
 
+// WithDefaultServiceConfig returns a DialOption that configures the default service config, which
+// will be used in cases where:
+// 1. WithDisableServiceConfig is called.
+// 2. Resolver does not return service config or if the resolver gets and invalid config.
+func WithDefaultServiceConfig(s string) DialOption {
```
Comment (jadekler) [inline dialoptions.go]: As mentioned over GVC, could we add a link or some documentation describing what this string should look like?

File: dialoptions.go [hunk 2]
```diff
@@ -446,6 +448,22 @@ func WithDisableServiceConfig() DialOption {
...
+	return newFuncDialOption(func(o *dialOptions) {
+		sc, err := parseServiceConfig(s)
+		if err != nil {
+			grpclog.Warningf("the provided service config is invalid, err: %v", err)
+			return
```
Comment (jadekler) [inline dialoptions.go]: This is unfortunate. Users won't find out until much later (if at all). Can we provide a ValidateServiceConfig option?

```
sc, err := grpc.ValidateServiceConfig("...")
if err != nil {
    panic(err)
}
opts = append(opts, grpc.WithDefaultServiceConfig(sc))
```

File: dialoptions.go (hunk 1 above)
Comment (lyuxuan) [reply dialoptions.go]: WIP to export a doc about canonical service config json string.

File: dialoptions.go (hunk 2 above)
Comment (lyuxuan) [reply dialoptions.go]: done

File: dialoptions.go (hunk 1 above)
Comment (jadekler) [reply dialoptions.go]: Ack

File: clientconn.go [hunk 3]
```diff
@@ -761,12 +767,36 @@ func (cc *ClientConn) getTransport(ctx context.Context, failfast bool, method st
...
 func (cc *ClientConn) handleServiceConfig(js string) error {
+	cc.mu.Lock()
+	defer cc.mu.Unlock()
 	if cc.dopts.disableServiceConfig {
-		return nil
+		if cc.dopts.defaultServiceConfig == nil {
+			return nil
+		}
+		// apply default service config when there's resolver service config is disabled.
+		return cc.applyServiceConfig(cc.dopts.defaultServiceConfig, cc.dopts.defaultServiceConfigRaw)
```
Comment (menghanl) [inline clientconn.go]: This will be called every time `handleServiceConfig` is called

File: clientconn.go [hunk 4]
```diff
@@ -761,12 +767,36 @@ func (cc *ClientConn) getTransport(ctx context.Context, failfast bool, method st
...
 		return nil
 	}
+
+	// the config has not been parsed, so we parse it now.
+	if sc == nil {
+		scfg, err := parseServiceConfig(js)
+		if err != nil {
+			return err
+		}
+		sc = &scfg
```
Comment (menghanl) [inline clientconn.go]: return pointer from `parseServiceConfig`?

File: dialoptions.go [hunk 5]
```diff
@@ -55,13 +55,15 @@ type dialOptions struct {
...
-	healthCheckFunc      internal.HealthChecker
+	resolverBuilder         resolver.Builder
+	reqHandshake            envconfig.RequireHandshakeSetting
+	channelzParentID        int64
+	disableServiceConfig    bool
+	disableRetry            bool
+	disableHealthCheck      bool
+	healthCheckFunc         internal.HealthChecker
+	defaultServiceConfig    *ServiceConfig
+	defaultServiceConfigRaw string
```
Comment (menghanl) [inline dialoptions.go]: Add a field to `ServiceConfig` for the raw js string.

File: dialoptions.go [hunk 6]
```diff
@@ -446,6 +448,22 @@ func WithDisableServiceConfig() DialOption {
 	})
 }
 
+// WithDefaultServiceConfig returns a DialOption that configures the default service config, which
+// will be used in cases where:
+// 1. WithDisableServiceConfig is called.
+// 2. Resolver does not return service config or if the resolver gets and invalid config.
+func WithDefaultServiceConfig(s string) DialOption {
+	return newFuncDialOption(func(o *dialOptions) {
+		sc, err := parseServiceConfig(s)
```
Comment (menghanl) [inline dialoptions.go]: We still give users a chance to pass invalid string here.

How about returning a dial option and an error from `ValidateServiceConfig()`, so the only way to create a default service config dial option is to validate it first?

The name `ValidateServiceConfig` would be inappropriate, time to bikeshed.

Comment (lyuxuan) [reply dialoptions.go]: If we have something like
```
func WithDefaultServiceConfig(s string) (DialOption, error) {
	sc, err := parseServiceConfig(s)
	if err != nil {
		return nil, fmt.Errorf("the provided service config is invalid, err: %v", err)
	}
	return newFuncDialOption(func(o *dialOptions) {
		o.defaultServiceConfig = sc
	}), nil
}
```
Then it breaks the idiomatic way of doing `grpc.Dial(target, grpc.WithXXX(), grpc.WithYYY())`. I think separating the validation is ok. I can add a comment saying that it's strongly recommended to verify the validity of the json string with the `ValidateServiceConfig` function.

File: clientconn.go (hunk 4 above)
Comment (lyuxuan) [reply clientconn.go]: done

File: dialoptions.go (hunk 5 above)
Comment (lyuxuan) [reply dialoptions.go]: done

File: dialoptions.go (hunk 6 above)
Comment (lyuxuan) [reply dialoptions.go]: Per offline discussion, we will check the validity of the default service config in `DialContext`. If it is not valid, the Dial will fail.

File: clientconn.go [hunk 7]
```diff
@@ -457,6 +467,24 @@ func (cc *ClientConn) waitForResolvedAddrs(ctx context.Context) error {
 	}
 }
 
+// gRPC should resort to default service config when:
+// * resolver service config is disabled
+// * or, resolver does not return a service config or returns an invalid one.
+func (cc *ClientConn) fallbackToDefaultServiceConfig(sc string) bool {
+	if cc.dopts.disableServiceConfig {
```
Comment (menghanl) [inline clientconn.go]: Update comment of `WithDisableServiceConfig`.

File: clientconn.go [hunk 8]
```diff
@@ -467,26 +495,18 @@ func (cc *ClientConn) updateResolverState(s resolver.State) error {
 		return nil
 	}
 
-	if !cc.dopts.disableServiceConfig && cc.scRaw != s.ServiceConfig {
-		// New service config; apply it.
+	if cc.fallbackToDefaultServiceConfig(s.ServiceConfig) {
+		if cc.dopts.defaultServiceConfig != nil && cc.sc.rawJSONString == nil {
```
Comment (menghanl) [inline clientconn.go]: Make `cc.sc` a pointer?

File: clientconn.go [hunk 9]
```diff
@@ -748,6 +768,28 @@ func (cc *ClientConn) getTransport(ctx context.Context, failfast bool, method st
 	return t, done, nil
 }
 
+func (cc *ClientConn) applyServiceConfig(sc *ServiceConfig) error {
```
Comment (menghanl) [inline clientconn.go]: Does the balancer get the default service config?

File: service_config.go [hunk 10]
```diff
@@ -99,6 +99,10 @@ type ServiceConfig struct {
 	// healthCheckConfig must be set as one of the requirement to enable LB channel
 	// health check.
 	healthCheckConfig *healthCheckConfig
+	// rawJSONString stores the pointer to the original service config json string that get parsed into
+	// this service config struct. Null pointer means this is a service config struct just defined,
+	// but no json string has been parsed and initialize the struct.
+	rawJSONString *string
```
Comment (menghanl) [inline service_config.go]: Does this still need to be a pointer?

**Pull request merged**